RecoTracker/TkSeedGenerator/plugins/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,380 @@
+    //building a tsos out of a VectorHit
+    const TrajectoryStateOnSurface initialTSOS = buildInitialTSOS(hitL3);
+    float signZ = copysign(1.0, initialTSOS.globalPosition().z());
+    float signPz = copysign(1.0, initialTSOS.globalMomentum().z());
+
+    //set the direction of the propagator
+    if (signZ * signPz > 0.0)
+      searchingPropagator->setPropagationDirection(oppositeToMomentum);
+    if (signZ * signPz < 0.0)
+      searchingPropagator->setPropagationDirection(alongMomentum);
+
+    //find vHits in layer 2
+    std::vector<TrajectoryMeasurement> measurementsL2 =
+        layerMeasurements_->measurements(*barrelOTLayer2, initialTSOS, *searchingPropagator, *estimator_);
+
+    std::vector<TrajectoryMeasurement>::iterator measurementsL2end =
+        std::remove_if(measurementsL2.begin(), measurementsL2.end(), isInvalid());
+    measurementsL2.erase(measurementsL2end, measurementsL2.end());
+
+    if (!measurementsL2.empty()) {
+      //not sure if building it everytime takes time/memory
+      const DetLayer* barrelOTLayer1 = measurementTracker_->geometricSearchTracker()->tobLayers().at(0);
+
+      for (const auto& mL2 : measurementsL2) {
+        const TrackingRecHit* hitL2 = mL2.recHit().get();
+
+        //propagate to the L2 and update the TSOS
+        std::pair<bool, TrajectoryStateOnSurface> updatedTSOS =
+            propagateAndUpdate(initialTSOS, *searchingPropagator, *hitL2);
+        if (!updatedTSOS.first)
+          continue;
+
+        //searching possible VHs in L1
+        std::vector<TrajectoryMeasurement> measurementsL1 =
+            layerMeasurements_->measurements(*barrelOTLayer1, updatedTSOS.second, *searchingPropagator, *estimator_);
+        std::vector<TrajectoryMeasurement>::iterator measurementsL1end =
+            std::remove_if(measurementsL1.begin(), measurementsL1.end(), isInvalid());
+        measurementsL1.erase(measurementsL1end, measurementsL1.end());
+
+        if (!measurementsL1.empty()) {
+          for (const auto& mL1 : measurementsL1) {
+            const TrackingRecHit* hitL1 = mL1.recHit().get();
+
+            //propagate to the L1 and update the TSOS
+            std::pair<bool, TrajectoryStateOnSurface> updatedTSOSL1 =
+                propagateAndUpdate(updatedTSOS.second, *searchingPropagator, *hitL1);
+            if (!updatedTSOSL1.first)
+              continue;
+
+            edm::OwnVector<TrackingRecHit> container;
+            container.push_back(hitL1->clone());
+            container.push_back(hitL2->clone());
+            container.push_back(hitL3.clone());
+
+            //building trajectory inside-out
+            if (searchingPropagator->propagationDirection() == alongMomentum) {
+              buildingPropagator->setPropagationDirection(oppositeToMomentum);
+            } else if (searchingPropagator->propagationDirection() == oppositeToMomentum) {
+              buildingPropagator->setPropagationDirection(alongMomentum);
+            }
```

Comment:
```suggestion
            } else {
              buildingPropagator->setPropagationDirection(alongMomentum);
            }
```
seems more appropriate

Reply:
Done.